agent/util-scripts/pbench-tool-meister

```diff
@@ -0,0 +1,998 @@
+    and the directory in which to store the data. In JSON form it will look
+    like:
+
+        {
+            "state":     "<'start'|'stop'|'send'>",
+            "group":     "<tool group name>",
+            "directory": "<directory in which to store tool data>"
+        }
+
+    If the Tool Meister is running on the same host as the pbench agent
+    controller, then the Tool Meister will write it's data directly to the
+    ${benchmark_results_dir} using the controller's host name; if the Tool
+    Meister is running remotely, then it will use a local temporary directory
+    to write it's data, and will send that data to the Tool Data Sink during
+    the "send" phase.
+    """
+
+    @staticmethod
+    def fetch_params(params):
+        """Static help method that allows the method constructing a ToolMeister
+        instance to verify the parameters before we actually construct the
+        object.
+
+        The definition of the contents of a parameter block is really
+        independent of a ToolMeister implementation, but we keep this method
+        in the ToolMeister class since it is closely related to the
+        implementation.
+        """
+        try:
+            benchmark_run_dir = params["benchmark_run_dir"]
+            channel = params["channel"]
+            controller = params["controller"]
+            group = params["group"]
+            hostname = params["hostname"]
+            tools = params["tools"]
+        except KeyError as exc:
+            raise ToolMeisterError(f"Invalid parameter block, missing key {exc}")
+        else:
+            return benchmark_run_dir, channel, controller, group, hostname, tools
+
+    def __init__(self, pbench_bin, params, redis_server, logger):
+        self.logger = logger
+        self.pbench_bin = pbench_bin
+        ret_val = self.fetch_params(params)
+        (
+            self._benchmark_run_dir,
+            self._channel,
+            self._controller,
+            self._group,
+            self._hostname,
+            self._tools,
+        ) = ret_val
+        self._running_tools = dict()
+        self._rs = redis_server
+        logger.debug("pubsub")
+        self._pubsub = self._rs.pubsub()
+        logger.debug("subscribe %s", self._channel)
+        self._pubsub.subscribe(self._channel)
+        logger.debug("listen")
+        self._chan = self._pubsub.listen()
+        logger.debug("done listening")
+        # Name of the temporary tool data directory to use when invoking
+        # tools.  This is a local temporary directory when the Tool Meister is
+        # remote from the pbench controller.
+        if self._controller != self._hostname:
+            self._tmp_dir = Path(
+                tempfile.mkdtemp(
+                    dir=os.environ.get("_PBENCH_TOOL_MEISTER_TMP", "/var/tmp"),
+                    prefix=f"tm.{self._group}.{os.getpid()}.",
+                )
+            )
+        else:
+            self._tmp_dir = None
```

Comment:
This self._tmp_dir construction is a bit awkward right in the middle of the redis setup; you listen, then go on to process it... but in the middle take a detour to set up a completely unrelated object property? It'd be easier to read if the _tmp_dir initialization were moved out of the sequence.

Reply:
I wouldn't mind having this fixed now if you have time; it would make reading easier. However, it's also not an important functional change and I know you'd like to get this in by Friday; so it can wait.

Reply:
Refactored this out in the tool-trigger changes that I just posted.